File: ios/Sources/Omnesis/DevAnnotations/DevAnnotationTarget.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

import Foundation

/// The entity a developer annotation points at — the iOS analogue of the
/// portal's `deriveDevTarget`. A detail view publishes its target into the
/// `AppStore` (`.devTarget(_:)`); shake-to-annotate reads whatever is current
/// and files a note against it. When no entity is in focus the composer files
/// a free-form `route` note instead, so a note can always be captured.
///
/// Pure `Foundation` (no UIKit/SwiftUI) so it compiles and unit-tests on the
/// sim-less macOS logic lane. The `type` strings mirror the gateway's
/// `DEV_ANNOTATION_TARGET_TYPES`.
public struct DevAnnotationTarget: Equatable, Sendable {
    public let type: String
    public let id: String?
    public let label: String

    public init(type: String, id: String?, label: String) {
        self.type = type
        self.id = id
        self.label = label
    }

    public static func document(_ id: String, label: String? = nil) -> DevAnnotationTarget {
        .init(type: "document", id: id, label: label ?? "Document \(id)")
    }

    public static func brief(_ id: String, label: String? = nil) -> DevAnnotationTarget {
        .init(type: "brief", id: id, label: label ?? "Brief \(id)")
    }

    public static func openLoop(_ id: String, label: String? = nil) -> DevAnnotationTarget {
        .init(type: "open_loop", id: id, label: label ?? "Loop \(id)")
    }

    public static func retiredLoop(_ id: String, label: String? = nil) -> DevAnnotationTarget {
        .init(type: "retired_loop", id: id, label: label ?? "Retired loop \(id)")
    }

    /// The singleton steward notes blob — no row id.
    public static func agentNotes(label: String = "Agent notes") -> DevAnnotationTarget {
        .init(type: "agent_notes", id: nil, label: label)
    }

    public static func agentRun(_ id: String, label: String? = nil) -> DevAnnotationTarget {
        .init(type: "agent_run", id: id, label: label ?? "Run \(id)")
    }

    public static func firing(_ id: String, label: String? = nil) -> DevAnnotationTarget {
        .init(type: "firing", id: id, label: label ?? "Firing \(id)")
    }

    public static func temporalAnnotation(_ id: String, label: String? = nil) -> DevAnnotationTarget {
        .init(type: "temporal_annotation", id: id, label: label ?? "Temporal annotation \(id)")
    }

    public static func conversation(_ id: String, label: String? = nil) -> DevAnnotationTarget {
        .init(type: "conversation", id: id, label: label ?? "Conversation \(id)")
    }

    /// Free-form note tagged with the current screen — the fallback when no
    /// addressable entity is in focus.
    public static func route(_ label: String) -> DevAnnotationTarget {
        .init(type: "route", id: nil, label: label)
    }
}

File: android/app/src/main/kotlin/dev/omnesis/android/ui/devannotate/DevAnnotationTarget.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

package dev.omnesis.android.ui.devannotate

/**
 * The entity a developer annotation points at — the Android analogue of the
 * iOS `DevAnnotationTarget` and the portal's `deriveDevTarget` result. A shake
 * files a note against whatever the operator is looking at; when no
 * addressable entity is in focus the composer files a free-form `route` note
 * instead, so a note can always be captured.
 *
 * Framework-free (no `android.*` imports) so it stays unit-testable on the
 * plain-JVM lane. The `targetType` strings mirror the gateway's
 * `DEV_ANNOTATION_TARGET_TYPES`.
 */
data class DevAnnotationTarget(
    val targetType: String,
    val targetId: String?,
    val label: String,
    val deepLink: String?,
)

/**
 * Derive a developer-annotation target from the current nav destination.
 *
 * TODO: only document routes and `agent` resolve to addressable entities today.
 * Privacy exchanges/approvals, watch detail, briefs, and people detail all
 * carry their entity IDs in nav arguments already — teach this function to
 * read them (and add the matching gateway target types where none exist, e.g.
 * watches) so those screens file attached notes instead of free-form ones.
 *
 * @param route the current destination's route *pattern* (e.g.
 *   `document/{id}`, `agent`), as `NavDestination.route` reports it — never a
 *   filled address, so a `deepLink` is only set when it names a real address
 *   (a filled document path, a placeholder-free route) and stays null
 *   otherwise. A templated pattern must never reach the store as a deep link.
 * @param documentId the decoded `id` argument of a `document/{id}`
 *   destination, when the route carries one.
 * @param agentSessionId the agent conversation currently on screen, when the
 *   agent surface holds one — mirrors the iOS conversation target and the
 *   portal's `activeConvoId`.
 */
fun devTargetForRoute(
    route: String?,
    documentId: String?,
    agentSessionId: String?,
): DevAnnotationTarget {
    if (route?.startsWith("document/") == true && !documentId.isNullOrBlank()) {
        return DevAnnotationTarget(
            targetType = "document",
            targetId = documentId,
            label = "Document $documentId",
            deepLink = "document/$documentId",
        )
    }
    if (route == "agent" && !agentSessionId.isNullOrBlank()) {
        return DevAnnotationTarget(
            targetType = "conversation",
            targetId = agentSessionId,
            label = "Conversation $agentSessionId",
            deepLink = route,
        )
    }
    // No addressable entity in focus — a free-form note tagged with the route.
    // The label keeps the pattern for screen identity; the deep link only a
    // placeholder-free address.
    return DevAnnotationTarget(
        targetType = "route",
        targetId = null,
        label = if (route.isNullOrBlank()) "General note" else "General note — $route",
        deepLink = route?.takeIf { it.isNotBlank() && !it.contains("{") },
    )
}
